lib/logger/lib/logger/backends/console.ex

```diff
@@ -19,15 +23,39 @@ defmodule Logger.Backends.Console do
     {:ok, :ok, configure(state.device, options)}
   end
 
+  def handle_call(:flush, state) do
+    {:ok, :ok, flush_buffer(state)}
+  end
+
   def handle_event({_level, gl, _event}, state)
   when node(gl) != node() do
     {:ok, state}
   end
 
   def handle_event({level, _gl, {Logger, msg, ts, md}}, state) do
-    if meet_level?(level, state.level) do
-      log_event(level, msg, ts, md, state)
+    %{level: log_level, ref: ref, batch_size: batch_size,
+      max_buffer: max_buffer} = state
+    cond do
+      not meet_level?(level, log_level) ->
+        {:ok, state}
+      is_nil(ref) ->
+        {:ok, log_event(level, msg, ts, md, state)}
+      batch_size < max_buffer ->
+        {:ok, buffer_event(level, msg, ts, md, state)}
+      true ->
+        {:ok, %{state | batch_size: batch_size + 1}}
     end
+  end
+
+  def handle_info({:io_reply, ref, :ok}, %{ref: ref} = state) do
+    {:ok, flush_buffer(%{state | ref: nil})}
+  end
+
+  def handle_info({:io_reply, ref, {:error, error}}, %{ref: ref} = state) do
```

Comment:
In my tests the error did not include the original message and it was simple `{:error, :put_chars}`. That's why I stored the previous message in the state.